torchgeo/datasets/openstreetmap.py

```diff
@@ -0,0 +1,555 @@
+
+        Args:
+            osm_data: JSON response from Overpass API
+
+        Returns:
+            GeoDataFrame containing parsed geometries, properties, and pre-computed labels
+        """
+        geometries = []
+        properties = []
+
+        for element in osm_data.get('elements', []):
+            geom = self._element_to_geometry(element)
+            if geom is not None:
+                geometries.append(geom)
+                props = element.get('tags', {}).copy()
+                props['osm_id'] = element.get('id')
+                props['osm_type'] = element.get('type')
+                properties.append(props)
+
+        if not geometries:
+            return gpd.GeoDataFrame(
+                columns=['geometry', 'label'], geometry='geometry', crs='EPSG:4326'
+            )
+
+        labels = [self._get_class_label({'properties': props}) for props in properties]
+
+        for props, label in zip(properties, labels):
+            props['label'] = label
+
+        gdf = gpd.GeoDataFrame(properties, geometry=geometries, crs='EPSG:4326')
+        return gdf
+
+    def _element_to_geometry(self, element: dict[str, Any]) -> shapely.Geometry | None:
+        """Convert OSM element to Shapely geometry.
+
+        Args:
+            element: OSM element from Overpass API response
+
+        Returns:
+            Shapely geometry or None if conversion fails
+        """
+        element_type = element.get('type')
+
+        with contextlib.suppress(KeyError, ValueError, TypeError):
+            if element_type == 'node':
+                lat = element.get('lat')
+                lon = element.get('lon')
+                if lat is not None and lon is not None:
+                    return shapely.Point(lon, lat)
+
+            elif element_type in ('way', 'relation'):
+                if 'geometry' in element:
+                    coords = [
+                        (node['lon'], node['lat']) for node in element['geometry']
+                    ]
+                    if len(coords) >= 2:
+                        tags = element.get('tags', {})
+                        is_area = tags.get('area') == 'yes' or (
+                            len(coords) >= 4 and coords[0] == coords[-1]
+                        )
+                        return (
+                            shapely.Polygon(coords)
+                            if is_area
+                            else shapely.LineString(coords)
+                        )
+
+        return None
+
+    def __getitem__(self, query: GeoSlice) -> dict[str, Any]:
+        """Retrieve input, target, and/or metadata indexed by spatiotemporal slice.
+
+        Args:
+            query: [xmin:xmax:xres, ymin:ymax:yres, tmin:tmax:tres] coordinates to index.
+
+        Returns:
+            Sample of input, target, and/or metadata at that index.
+
+        Raises:
+            IndexError: If *query* is not found in the index.
+        """
+        self._ensure_empty_classes_checked()
```

Comment:
So this method is called on the entire dataset, not just the query slice. Should this be in `__init__` instead of `__getitem__`?